File: coding/src/com/algos/core/li03_slidingwindow/revised/L1004_MaxConsecutiveOnesIII.java
package com.algos.core.li03_slidingwindow.revised;

public class L1004_MaxConsecutiveOnesIII {
    
        public static void main(String[] args) {
            L1004_MaxConsecutiveOnesIII obj = new L1004_MaxConsecutiveOnesIII();
    
            int[] nums = {1,1,1,0,0,0,1,1,1,1,0};
            int k = 2;
            System.out.println(obj.longestOnes(nums, k));
        }
    
        // Problem: https://leetcode.com/problems/max-consecutive-ones-iii/
        // Idea: This can be reframed as find the longest subarray with at most k zeros
        // Time: O(n)
        public int longestOnes(int[] nums, int k) {
            int l=0, r=0, n = nums.length, maxLen = 0, zeros = 0;
            while(r<n) {
                if(nums[r] == 0) 
                    zeros++;
                
                if(zeros > k) {
                    if(nums[l] == 0) 
                        zeros--;
                    l++;
                }
                
                if(zeros <= k)
                    maxLen = Math.max(maxLen, r-l+1);
                r++;
            }
            
            return maxLen;
        }
}

File: dsa/patterns/04_slidingwindow/python/l1004_max_consecutive_ones_iii.py
"""
L1004 - Max Consecutive Ones III

Problem: https://leetcode.com/problems/max-consecutive-ones-iii/
Idea: This can be reframed as find the longest subarray with at most k zeros
Time: O(n)
Space: O(1)
"""

def longest_ones(nums, k):
    l, r, n, max_len, zeros = 0, 0, len(nums), 0, 0
    
    while r < n:
        if nums[r] == 0:
            zeros += 1
        
        if zeros > k:
            if nums[l] == 0:
                zeros -= 1
            l += 1
        
        if zeros <= k:
            max_len = max(max_len, r - l + 1)
        
        r += 1
    
    return max_len


if __name__ == "__main__":
    nums = [1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0]
    k = 2
    print(f"Input: nums = {nums}, k = {k}")
    print(f"Output: {longest_ones(nums, k)}")
